### DISCUSSION TITLE
Method To Make Use of Zim's Formatting Menu Within A Plugin Buffer?

### DISCUSSION THREAD

**Discussion by mbluett88** [zim-desktop-wiki/zim-desktop-wiki#2566] [Development]
I am trying to find a method of causing the Zim formatting options to be applied to the formatting within a plugin buffer. Does such a method or process exist within Zim?

For example, if I want to click on the Zim bold menu icon and apply that formatting to some selected text within a plugin buffer, how would I go about doing that?

**Comment by jaap-karssenberg** [comment]
Afraid at the moment there is no standard wat to do that.

Is the buffet inside an embedded object in the main buffer, or part of some
dialog etc. ?

For nested buffers I do have a design ideas how to make that work but not
yet implemented.

For a dialog you would re-use the relevant widgets but you have to hook
them up yourself.

Regards,

Jaap


Op do 1 feb. 2024 21:57 schreef mbluett88 ***@***.***>:

**Reply by mbluett88** [reply]
> Is the buffet inside an embedded object in the main buffer, or part of some
dialog etc. ?

It is an embedded buffer as I am using InsertedObjectTypeExtension. This is based on a text expander section (Gtk.Expander) that is an object within a normal Zim page (like that of CodeView) except that it requires the same formatting that Zim uses.

In a discussion with Introt, he/she suggested that I try an incorporate the Expander object directly into the Zim base code. Maybe that is the optimal way to do it. What is your feeling?

> For a dialog you would re-use the relevant widgets but you have to hook them up yourself.

I tried the following but I can't get it to work:

`@_install_format_actions
class ExpanderWidget(InsertedObjectWidget):

	def __init__(self, plugin, notebook, page, textbuffer):
		InsertedObjectWidget.__init__(self)
		assert Gio is not None
		self._gaction = Gio.SimpleAction.new_stateful('apply_format_strong', None, GLib.Variant.new_boolean(True))
		self._gaction.set_enabled(True)
		self._gaction.connect('activate', self.do_toggle_format_action)

I copied the `do_toggle_format_action` and `toggle_format` from PageView and modified `toggle_format` to make use of my TextView.

Regardless of this `connect` being present Zim only makes use of the one PageView knows about.

So, I am basically at an impasse. Unless, you can suggest an alternate method of implementation.

Maybe the route to go is to incorporate the Expander object directly into PageView as was suggested by Introt.

**Reply by mbluett88** [reply]
The reason it only calls PageView's `do_toggle_format_action` is because of the following:

```
class ToggleActionMethod(ActionMethod):

	def __call__(self, active=None):
		if not self._sensitive:
			raise AssertionError('Action not sensitive: %s' % self.name)

		with self._on_activate.blocked():
			# This results in PageView: do_toggle_format_action_alt() method being called.
			self._action.func(self._instance, active)

```
Unless, I modify this code or add the Expander object directly into PageView I see no way around this problem.

**Comment by introt** [comment]
See also: #818 (Refactor table editor to use nested textview instead of treeview) - the proposed "nested textview" that'd supposedly live inside each cell(?) could also work for your text expanders
